sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala

```diff
@@ -134,3 +139,87 @@ object ParserUtils extends SparkParserUtils {
     sb.toString()
   }
 }
+
+object LabelUtils {
```

Comment:
is this class threadsafe? Users may invoke the parser in multiple threads. At least we should put states in ThreadLocal.

Reply:
Or we pass around a `LabelUtils` instance when parsing a single query.

Reply:
This crossed my mind yesterday, but we weren't sure. If that's the case, then you are right - I would say we just need to make `seenLabels` thread local because that's the only state required to track the labels.

Reply:
@cloud-fan is `AstBuilder` instantiated every time when parser is invoked? Is it safe to keep some state inside `AstBuilder` class?

For `LabelUtils` for now is it enough to make `seenLabels` ThreadLocal? Can single parsing invocation be done from different threads, like threadpool with different threads that get appointed to do part of the parsing?

Reply:
I got it, thanks!